src/Init/System/IO.lean

```diff
@@ -97,48 +175,39 @@ def sleep (ms : UInt32) : IO Unit :=
   -- TODO: add a proper primitive for IO.sleep
   fun s => dbgSleep ms fun _ => EStateM.Result.ok () s
 
-/--
-  Run `act` in a separate `Task`. This is similar to Haskell's [`unsafeInterleaveIO`](http://hackage.haskell.org/package/base-4.14.0.0/docs/System-IO-Unsafe.html#v:unsafeInterleaveIO),
-  except that the `Task` is started eagerly as usual. Thus pure accesses to the `Task` do not influence the impure `act`
-  computation.
-  Unlike with pure tasks created by `Task.mk`, tasks created by this function will be run even if the last reference
-  to the task is dropped. `act` should manually check for cancellation via `IO.checkCanceled` if it wants to react
-  to that. -/
-@[extern "lean_io_as_task"]
-constant asTask (act : IO α) (prio := Task.Priority.default) : IO (Task (Except IO.Error α))
+/-- `IO` specialization of `EIO.asTask`. -/
+@[inline] def asTask (self : IO α) (prio := Task.Priority.default) : RealM (Task (Except IO.Error α)) :=
+  EIO.asTask self prio
 
-/-- See `IO.asTask`. -/
-@[extern "lean_io_map_task"]
-constant mapTask (f : α → IO β) (t : Task α) (prio := Task.Priority.default) : IO (Task (Except IO.Error β))
+/-- `IO` specialization of `EIO.mapTask`. -/
+@[inline] def mapTask (f : α → IO β) (t : Task α) (prio := Task.Priority.default) : RealM (Task (Except IO.Error β)) :=
+  EIO.mapTask f t prio
 
-/-- See `IO.asTask`. -/
-@[extern "lean_io_bind_task"]
-constant bindTask (t : Task α) (f : α → IO (Task (Except IO.Error β))) (prio := Task.Priority.default) : IO (Task (Except IO.Error β))
+/-- `IO` specialization of `EIO.bindTask`. -/
+@[inline] def bindTask (t : Task α) (f : α → IO (Task (Except IO.Error β))) (prio := Task.Priority.default) : RealM (Task (Except IO.Error β)) :=
+  EIO.bindTask t f prio
 
-def mapTasks (f : List α → IO β) (tasks : List (Task α)) (prio := Task.Priority.default) : IO (Task (Except IO.Error β)) :=
-  go tasks []
-where
-  go
-    | t::ts, as =>
-      IO.bindTask t (fun a => go ts (a :: as)) prio
-    | [], as => IO.asTask (f as.reverse) prio
+/-- `IO` specialization of `EIO.mapTasks`. -/
+@[inline] def mapTasks (f : List α → IO β) (tasks : List (Task α)) (prio := Task.Priority.default) : RealM (Task (Except IO.Error β)) :=
+  EIO.mapTasks f tasks prio
 
 /-- Check if the task's cancellation flag has been set by calling `IO.cancel` or dropping the last reference to the task. -/
-@[extern "lean_io_check_canceled"] constant checkCanceled : IO Bool
+@[extern "lean_io_check_canceled"] constant checkCanceled : EIO ε Bool
```

Comment:
Changing this from `EIO Empty` can't seriously be the solution to the lifting issue, can it?

Reply:
I didn't do this elsewhere because this style, on the other hand, prevents automatic lifting of the monad in some cases. As such, there is a where negotiation between where `EIO e` provides better type inference and where `RealM` does.

Reply:
I believe we can use `BaseIO` now without breaking the test

Reply:
Ah, I should have unresolved this one

Reply:
> I believe we can use `BaseIO` now without breaking the test

@Kha I am 99% we cannot, as I had previously tried to do so and it failed. But I will test again just to make sure.

Reply:
Well, it appears I am wrong, it does now work!

Reply:
As such, I am going to change `getNumHearbeats` as well.